### PR TITLE
[JFG-979] Expand java based configuration with listeners support

### PR DESCRIPTION
**Pull request by a-badaev**

### CODE REVIEW

File: chassis/core/src/main/java/com/griddynamics/jagger/user/test/configurations/JTestDefinition.java [hunk 1]
```diff
@@ -111,19 +112,68 @@ public Builder withInvoker(Class<? extends Invoker> invoker) {
...
-        public Builder withValidators(List<Class<? extends ResponseValidator>> validators) {
-            this.validators = validators;
+        public Builder addValidators(List<Class<? extends ResponseValidator>> validators) {
+            this.validators.addAll(validators);
             return this;
         }
-
+    
+        /**
+         * Optional: Adds a subtypes of {@link com.griddynamics.jagger.engine.e1.collector.ResponseValidator}
```
Comment (dlatnikov) [inline chassis/core/src/main/java/com/griddynamics/jagger/user/test/configurations/JTestDefinition.java]: a subtype__

File: chassis/core/src/main/java/com/griddynamics/jagger/util/generators/TestGenerator.java [hunk 2]
```diff
@@ -23,10 +23,12 @@ static WorkloadTask generateFromTest(JLoadTest jLoadTest,
         WorkloadTask task = generatePrototype(jLoadTest.getTestDescription());
         task.setName(jLoadTest.getId());
         task.setVersion("0");
-        TerminateStrategyConfiguration terminateStrategyConfiguration = generateTermination(jLoadTest.getTermination());
-        task.setTerminateStrategyConfiguration(terminateStrategyConfiguration);
-        WorkloadClockConfiguration workloadClockConfiguration = generateLoad(jLoadTest.getLoad());
-        task.setClockConfiguration(workloadClockConfiguration);
+        task.setTestListeners(jLoadTest.getTestListeners());
```
Comment (dlatnikov) [inline chassis/core/src/main/java/com/griddynamics/jagger/util/generators/TestGenerator.java]: add standard CollectThreadsTestListener to the list as well, you have removed it in the generatePrototype()

File: chassis/core/src/test/java/com/griddynamics/jagger/ExampleJLoadScenarioProvider.java [hunk 3]
```diff
@@ -15,34 +27,53 @@
...
+                .withWarmUpTimeInSeconds(10)
+                .build();
         
-        JTerminationCriteria jTerminationCriteria = JTerminationCriteriaIterations.of(IterationsNumber.of(1000), MaxDurationInSeconds.of(20));
+        JTerminationCriteria jTerminationCriteria = JTerminationCriteriaIterations
+                .of(IterationsNumber.of(1000), MaxDurationInSeconds.of(20));
+    
+        JLimit jLimit = JLimitVsRefValue
+                .builder("throughput", RefValue.of(4d))
+                .withOnlyLowerThresholds(LowWarnThresh.of(1d), LowErrThresh.of(2.d))
```
Comment (dlatnikov) [inline chassis/core/src/test/java/com/griddynamics/jagger/ExampleJLoadScenarioProvider.java]: incorrect

File: archetype-java-builders/src/main/resources/archetype-resources/src/main/java/ExampleJLoadScenarioProvider.java [hunk 4]
```diff
@@ -26,21 +37,40 @@ public static JLoadScenario getExampleJaggerLoadScenario() {
...
                 .build();
         
-        JLoadProfile jLoadProfileRps = JLoadProfileRps.builder(RequestsPerSecond.of(10)).withMaxLoadThreads(10).withWarmUpTimeInSeconds(10).build();
+        JTerminationCriteria jTerminationCriteria = JTerminationCriteriaIterations
+                .of(IterationsNumber.of(1000), MaxDurationInSeconds.of(20));
         
-        JTerminationCriteria jTerminationCriteria = JTerminationCriteriaIterations.of(IterationsNumber.of(1000), MaxDurationInSeconds.of(20));
+        JLimit jLimit = JLimitVsRefValue
+                .builder("throughput", RefValue.of(4d))
+                .withOnlyLowerThresholds(LowWarnThresh.of(1d), LowErrThresh.of(2.d))
```
Comment (a-badaev) [inline archetype-java-builders/src/main/resources/archetype-resources/src/main/java/ExampleJLoadScenarioProvider.java]: how it should be?

File: chassis/core/src/main/java/com/griddynamics/jagger/util/generators/TestGenerator.java (hunk 2 above)
Comment (a-badaev) [inline chassis/core/src/main/java/com/griddynamics/jagger/util/generators/TestGenerator.java]: you mean in adddition to user specified ones it must be always added?

File: chassis/core/src/test/java/com/griddynamics/jagger/ExampleJLoadScenarioProvider.java (hunk 3 above)
Comment (a-badaev) [inline chassis/core/src/test/java/com/griddynamics/jagger/ExampleJLoadScenarioProvider.java]: how it should be ?

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/collector/testsuite/ExampleTestSuiteListener.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    
+    private final static Logger log = LoggerFactory.getLogger(ExampleTestSuiteListener.class);
+    
+    @Override
+    protected void init() {
+        super.init();
+    }
+    
+    @Override
+    public TestSuiteListener provide() {
```
Comment (dlatnikov) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/collector/testsuite/ExampleTestSuiteListener.java]: yep

File: chassis/core/src/main/java/com/griddynamics/jagger/util/generators/TestGenerator.java (hunk 2 above)
Comment (dlatnikov) [inline chassis/core/src/main/java/com/griddynamics/jagger/util/generators/TestGenerator.java]: yes

File: archetype-java-builders/src/main/resources/archetype-resources/src/main/java/ExampleJLoadScenarioProvider.java (hunk 4 above)
Comment (dlatnikov) [inline archetype-java-builders/src/main/resources/archetype-resources/src/main/java/ExampleJLoadScenarioProvider.java]: just remove it for now

File: chassis/core/src/main/java/com/griddynamics/jagger/user/test/configurations/JLoadScenario.java [hunk 6]
```diff
@@ -29,11 +35,13 @@
     private final String id;
     private final List<JParallelTestsGroup> testGroups;
     private final List<Double> percentileValues;
+    private final List<Provider<TestSuiteListener>> listeners;
```
Comment (dlatnikov) [inline chassis/core/src/main/java/com/griddynamics/jagger/user/test/configurations/JLoadScenario.java]: why is it null here, but has empty array value in other cases?

File: chassis/core/src/main/java/com/griddynamics/jagger/user/test/configurations/JLoadTest.java [hunk 7]
```diff
@@ -32,18 +39,19 @@
     private final JTestDefinition testDescription;
     private final JTerminationCriteria termination;
     private final List<JLimit> limits;
+    private final List<Provider<TestListener>> testListeners;
```
Comment (dlatnikov) [inline chassis/core/src/main/java/com/griddynamics/jagger/user/test/configurations/JLoadTest.java]: let's rename to listeners and also getter

File: chassis/core/src/main/java/com/griddynamics/jagger/user/test/configurations/JLoadScenario.java (hunk 6 above)
Comment (a-badaev) [inline chassis/core/src/main/java/com/griddynamics/jagger/user/test/configurations/JLoadScenario.java]: it is like everywhere. in the builder it is an empty array

File: chassis/core/src/main/java/com/griddynamics/jagger/user/test/configurations/JLoadTest.java (hunk 7 above)
Comment (a-badaev) [inline chassis/core/src/main/java/com/griddynamics/jagger/user/test/configurations/JLoadTest.java]: done

**Pull request merged**